src/mongoc/mongoc-write-command.c

```diff
@@ -706,18 +781,65 @@ _mongoc_write_command_update (mongoc_write_command_t       *command,
       EXIT;
    }
 
+   if ((command->u.update.n_updates == 0) ||
+       !bson_iter_init (&iter, command->u.update.updates) ||
+       !bson_iter_next (&iter)) {
+      bson_set_error (error,
+                      MONGOC_ERROR_COLLECTION,
+                      MONGOC_ERROR_COLLECTION_UPDATE_FAILED,
+                      "Cannot do an empty update.");
+      result->failed = true;
+      EXIT;
+   }
+
+   overhead = 1 + strlen ("updates") + 1;
+
+again:
+   bson_init (&cmd);
+   has_more = false;
+   i = 0;
+
    BSON_APPEND_UTF8 (&cmd, "update", collection);
    BSON_APPEND_DOCUMENT (&cmd, "writeConcern",
                          WRITE_CONCERN_DOC (write_concern));
-   BSON_APPEND_BOOL (&cmd, "ordered", command->u.update.ordered);
-   bson_append_array_begin (&cmd, "updates", 7, &ar);
-   bson_append_document_begin (&ar, "0", 1, &child);
-   BSON_APPEND_DOCUMENT (&child, "q", command->u.update.selector);
-   BSON_APPEND_DOCUMENT (&child, "u", command->u.update.update);
-   BSON_APPEND_BOOL (&child, "multi", command->u.update.multi);
-   BSON_APPEND_BOOL (&child, "upsert", command->u.update.upsert);
-   bson_append_document_end (&ar, &child);
-   bson_append_array_end (&cmd, &ar);
+   BSON_APPEND_BOOL (&cmd, "ordered", command->u.insert.ordered);
+
+   if ((command->u.update.updates->len < client->cluster.max_bson_size) &&
+       (command->u.update.updates->len < client->cluster.max_msg_size) &&
+       (command->u.update.n_updates <= max_update_batch)) {
+       // there is enough space to send the whole query at once...
+      BSON_APPEND_ARRAY (&cmd, "updates", command->u.update.updates);
+   } else {
+      bson_append_array_begin (&cmd, "updates", strlen("updates"), &ar);
```

Comment:
This is really ugly. We first built the updates and them split them up again. But this is the same behaviour as insert. 

It might make much more sense to split it up in mongoc_bulk_update_append. I tried this but for some reason the information about nodes might not yet exist in the client.

In general, this codebase has so many codepaths. Maybe try removing some of the deprecated stuff in the future?